core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```
@@ -270,6 +271,33 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     currentStateTimestamp = Some(time.milliseconds())
   }
 
+  /**
+    * Shrinks the consumer group to the given size by removing members while maintaining the leader
+    */
+  def shrinkTo(maxSize: Int): Unit = {
+    if (members.size < maxSize)
+      throw new IllegalArgumentException(s"Cannot shrink group $groupId to $maxSize as it's current count ${members.size} is smaller.")
```

do we really want to throw? Rather than just silently do nothing?

---

I don't have a strong opinion about this

---

Perhaps a log message (warning) would suffice?

---

Should we just use `assert` here? As long as the change is robust, this should never happen which makes me believe adding assertion will be a better approach to avoid future changes affecting group max size logic.

---

The argument name `maxSize` suggests this should be a no-op if the group is already smaller. I'd just remove the check.